Topics/System_Design/Twitter_System_Design.md

```diff
@@ -0,0 +1,63 @@
+# Twitter Case Study: How to Design the Home Timeline for Massive Scale
+
+## Introduction
+Twitter's home timeline is in theory a very simple feature. Users are in a many-to-many relationship with other users. Users can post Tweets. Each follower of the user should see every Tweet posted by the people they follow in their home feed, sorted in chronological order.
+
+The challenge for Twitter is building a highly-available and responsive home timeline given their massive scale. As of November 2012, Twitter had:
+- 150M world wide active users
+- Over 300K Queries-per-Second (QPS) for timelines
+
+In conjunction with this scale, Raffi Krikorian notes in his [QCon Talk](https://www.infoq.com/presentations/Twitter-Timeline-Scalability/) that when a user tweets, their followers should be updated with the new tweet in 5 seconds or under. This article will explain why this is a non-trivial task and give a high-level overview of how Twitter designed their systems to accomodate their scale (as of 2012).
+
+## Approach 1: Naive SQL Query
+Below is a simplified SQL schema of Twitter's data model.
+
+**tweets**
+| id   |      author_id      | content  |
+|----------|:-------------:|------:|
+| 38927 | 1  | "Did you ever hear the tragedy of Darth Plagueis The Wise? I thought not." |
+| 38928 |    2   |   "I hate sand." |
+| 38929 | 3 |    "You were my brother Anakin!" |
+
+**users**
+| id   |      username      |  bio |
+|----------|:-------------:|------:|
+| 1 |  darth-palpatine | Unlimited Power! |
+| 2 |    anakin-skywalker   |   The Jedi are Evil |
+| 3 | obi-wan |  Hello there   |
+
+**follows**
+| follower_id   |      follows_id      |
+|----------|:-------------:|
+| 2 |  1 |
+| 2 |    3   |
+| 1 | 2 |
+
+Twitter's initial naive approach was to join the Tweet, User, and Follows tables in SQL upon a GET request from a client like so:
+```
+SELECT * FROM tweets
+JOIN users ON tweets.author_id = users.id
+JOIN follows ON follows_id = users.id
+WHERE follower_id = current_user.id;
+```
+
+You can quickly see how this approach became quite slow at over 300K QPS.
+
+## Approach 2: Caching
+As a result, Twitter implemented a fan-out caching system. They stored a [Redis list](https://redis.io/docs/data-types/lists/) for each user containing the Tweets to display on each user's home timeline. Let's illustrate this with an example. Say that Darth Palpatine is followed by Anakin Skywalker, Obi-Wan, and Padme Amidala. When Darth Palpatine posts a Tweet, Twitter's fan-out service pushes the new Tweet to the Redis cache for each user that follows Darth Palpatine (Anakin, Obi-Wan, and Padme). This way, when Anakin (for example) loads his timeline page, his Tweets are cached in Redis and can be retrieved in constant runtime (in this simplified example).
```

Comment:
did they use Redis lists for a reason vs. other data structures or libraries? This isn't necessarily an issue with the write up here just curious / maybe worth comparing other possible approaches to elucidate why this is the most logical solution. perhaps could briefly examine other possible solutions?

Reply:
maybe also a quick blurb on what Redis is for those unfamiliar

Reply:
I think any in-depth discussion on Redis is out of the scope of the article, but I can add a short explanation on what Redis is, as well as a link to an article with some reasons that Twitter chose Redis.